### DISCUSSION TITLE
feat(Tab): Validation tooltips appear when not supposed to

### DISCUSSION THREAD

**Discussion by celadaris** [dotnetcore/BootstrapBlazor#4368] [Q&A] [answered]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When alternating the tabs using `<Tab IsOnlyRenderActiveTab="true">,` sometimes i will see that the behavour of `ShowAllInvalidResult="false"` breaks, and the tooltips appear after calling `Validate()` on some of the `ValidateForm`'s.

### Expected Behavior

Tooltips should not appear after `Validate()` has been called, because the `Validate()` behaviour should be consistent with the value of `ShowAllInvalidResult`.

### Interactive render mode

Interactive Server (Interactive server-side rendering (interactive SSR) using Blazor Server)

### Steps To Reproduce

- In the project i have a `<Tab IsOnlyRenderActiveTab="true">` and in the first tab i have a child component with many `ValidateForm` that have the default behavour of `ShowAllInvalidResult="false"` 

- I also have a couple different kinds of inputs within the `ValidateForm` each binded to a `[Required]` string.

- Within the `OnAfterRender` I want to `Validate()` each of the forms so that they validate when the child component initially loads.

Project:
[ValidFormsBugTest.zip](https://github.com/user-attachments/files/17091304/ValidFormsBugTest.zip)

Video Demonstration:

https://github.com/user-attachments/assets/b726dd1c-2a53-462b-8e9b-747a7f6399dd

### .NET Version

.net 8

**Comment by bb-auto** [comment]
@celadaris Thank you for filing this issue. In order for us to investigate this issue, please provide [a minimalistic repro project](https://github.com/dotnetcore/bootstrapblazor/blob/main/docs/repro.md) that illustrates the problem.

**Reply by ArgoZhang** [reply]
@celadaris 

It seems that this function is a bit difficult. the `ValidateForm` components have no relationship with each other. The parameter `ShowAllInvalidResult` value is responsible for the current form behavior. As shown in the reproduction project, multiple forms, each form only shows its own error prompt, which is correct, but this is not what you want. So this requires a new mechanism to uniformly manage all forms in the page


### Assume a scenario

1. put one submit button in the top 2 `ValidateForm`
2. switch to `tab1` from `tab2` (trigger all from validate)
3. click the submit button in the second form

### expected (maybe)
1. the first invalid element show tooltip in the top 1 form. the other form do not show tooltip 
2. the first invalid element in the second form show tooltip.

what to do?

**Reply by celadaris** [reply]
> As shown in the reproduction project, multiple forms, each form only shows its own error prompt, which is correct, but this is not what you want.

this is correct and this is what I originally wanted.

> this requires a new mechanism to uniformly manage all forms in the page

I think you are correct to think that this type of mechanism is needed. I thought `SectionOutlet `could theoretically be this mechanism, but it too has consistency problems when being used in a Tab with `IsOnlyRenderActiveTab="true"`

[OutletExample.zip](https://github.com/user-attachments/files/17164862/OutletExample.zip)

https://github.com/user-attachments/assets/f33134b9-ce44-454f-b321-49d6ca42b14a

**Reply by celadaris** [reply]
in your Scenario I had trouble understanding exactly what is happening, so i created 4 different scenarios of what i expect to consistently happen when we may or may not validate OnAfterRender and where we may or may not have ShowAllInvalidResult set to true,

[here](https://lucid.app/lucidchart/fb09157e-cd52-4344-ad09-6759f8ba6541/edit?viewport_loc=-124%2C383%2C3003%2C2666%2C0_0&invitationId=inv_af448d43-5d24-4890-a0cd-d5422b7e98c2) is a link to where I made the flowchart, you can edit it if you want, or just look at the image.

![Flowchart](https://github.com/user-attachments/assets/c73f2f24-87c8-4e42-b5ec-843bbd2275f8)

**Reply by celadaris** [reply]
This diagram demonstrates that if we choose to validate when loading then we have the ability to:

- load with tooltips, and have tooltips appear on submit
- load WITHOUT tooltips, and tooltips DONT appear on submit
- load WITHOUT tooltips, and have tooltips appear on submit

but we lack the functionality to:

- load with tooltips, and NOT have tooltips appear on submit

I personally don't have a need for the missing functionality, `<SectionOutlet>` could theoretically be used to add the missing functionality.

> So this requires a new mechanism to uniformly manage all forms in the page

Something like a Tag specifically for validating any `<ValidateForm>` inside of it when the component starts with an optional tooltip visibility parameter would be really nice, which i think is what you were suggesting. It would also provide the missing functionality.

some other nice to have parameters:
- some kind of event(s) to subscribe to when it triggers would be nice.
- parameter to choose whether to trigger OnValid/OnInvalid events in the ValidateForm's that it validates.

**Reply by ArgoZhang** [reply]
@celadaris hi bro. I made some updates to the code and update the mini-repro-project. Could you please help test it? Please check if the following meets your needs. Please refer the branch `feat-validate-form`

[ValidFormsBugTest-update.zip](https://github.com/user-attachments/files/17178435/ValidFormsBugTest-update.zip)

**Reply by celadaris** [reply]
This works perfectly for my needs, i do see that in either scenario where we do or dont validate `OnAfterRender()` and we have `ShowAllInvalidResult = true` the tooltips don't always appear on submit after the component loads. This does not affect me and my project but could affect others.

so in Scenario 1 and 3 is where we have consistency problems.

- In Scenario 1 they appear on validate perfectly fine, and when we press submit the tooltips stay visible and its consistent, but when we hover over them and make them dissapear then hit submit, we have consistency problems.

- In Scenario 3 On the first submit, the tooltips become visible as they should consistently, but when we hover over them and make them dissapear then hit submit, we have consistency problems.

[Pages.zip](https://github.com/user-attachments/files/17180174/Pages.zip)

https://github.com/user-attachments/assets/1a6b4a99-6892-4094-b002-24d3790ebcd6

**Reply by ArgoZhang** [reply]
@celadaris 


https://github.com/user-attachments/assets/62b17a8c-a594-4437-80f7-54a9c6ae0235

Try clearing the cache.

**Reply by celadaris** [reply]
OHHHHHHHHHHHHHH I FIGURED IT OUT!!! haha

**Comment by celadaris** [comment]
if you type `sysdm.cpl` in Run and under performance click on settings, then change your settings to Adjust for best appearance then you will see the bugs, and maybe you remember, but i had the problem a while ago where i had the null error in my console, This occurs aswell from one of these settings being different. This is demonstrated in the video aswell.


https://github.com/user-attachments/assets/363afa22-0277-40e6-a1bd-fd42500cb123

**Reply by ArgoZhang** [reply]
@celadaris Yes. Great job!!!

This is a known issue because the bootstrap style will enable animations based on system settings. this problem is very easy to reproduce caused by you turned on the animation.

```
@media (prefers-reduced-motion: reduce)
```

This problem troubles almost all users.

https://github.com/twbs/bootstrap/issues/37474?wt.mc_id=DT-MVP-5004174

Like I said, I tried many things to get rid of this error, but nothing worked.

the good news the `bs` team have a PR for this issue
https://github.com/twbs/bootstrap/pull/40856?wt.mc_id=DT-MVP-5004174

the bad news is this fix has not yet been merged into the master branch.

**Reply by celadaris** [reply]
Oh okay! Thank you!! Then from my testing on the tooltips things work as they should 🥳

**Reply by ArgoZhang** [reply, marked as answer]
@celadaris Ok. I will bump a beta version later.

this script error needs to wait for `bootstrap` to release a new version. give them some time.
